### DISCUSSION TITLE
ERMS, FSRM, dynamic SIMD vector lengths, and other memcpy() accelerators

### DISCUSSION THREAD

**Discussion by sh1boot** [zlib-ng/zlib-ng#1777] [General]
Discussion over [here](https://github.com/zlib-ng/zlib-ng/pull/1568#issuecomment-1792964528) (from approximately 3rd of November 2023 onwards) has gone much deeper into hardware `memcpy()` acceleration than is warranted under an old PR; so I propse to move the discussion here instead.

In essence, a large part of zlib's time is spent in memcpy() operations of varying lengths and there's a great deal of complexity in exploiting cases where the output is not sensitive to overwrite.  Modern hardware is not like the hardware that was around when those optimisations were designed (though there's plenty of that hardware still around).

At some point we should expect to see FSRM and variable-length vector operations reach a point where they're reliably faster than that complexity, and it's going to affect the architecture.

**Comment by sh1boot** [comment]
I also feel that there's a gaping hole in Clang's efforts to optimise known-short-but-not-constant memcpy operations.  It goes all the way down to libc, which imposes huge overheads if done too frequently.

**Comment by KungFuJesus** [comment]
If there's any gain I suspect it will only be for very recent Intel CPUs and it probably will be at least partially mitigated by recent security related microcode updates. I also don't think any of them cope very well with overlap side effects but I need an icelake CPU to verify that

**Comment by sh1boot** [comment]
I expect overlap to be hopelessly slow but correct, so that'll need special handling still. 

It may turn out to take longer than hoped for anybody to implement it well.

But that's fine because it's just that legacy x86-derived architecture which nobody cares about anymore.  There are more sensible architectures to look at as well.

**Comment by KungFuJesus** [comment]
So after some careful experiments to try to minimize the code a ton for basically all cases (I simply compute overlap, if it exists, and then in a loop call the FSRM sequence with either the distance between the pointers or the complete length of the copy), I found that this is in fact slower for everything with FSRM, which is funny.

I might play with it a little more but yeah, at best we wind up with something that is not punitively worse in my testing (on an i7-11800H).  It does make for a delightfully shorter code size when compiled, though.
